### PR TITLE
OpInfo & test_torch.py cleanup

### PR DESCRIPTION
**Pull request by imaginary-person**
### SUMMARY

Will repurpose this PR to cleanup `test_torch.py` & OpInfos.

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -549,21 +549,42 @@ def sample_inputs_linalg_vector_norm(op_info, device, dtype, requires_grad):
...
+        ((1,), (S, S), (S, M)),
+        ((), (S, S), (S, M))
+    )
+    inputs = list(SampleInput(make_tensor(shape_a, device, dtype, requires_grad=requires_grad),
+                              args=(make_tensor(shape_b, device, dtype,
+                                                requires_grad=requires_grad),
+                                    make_tensor(shape_c, device, dtype,
+                                                requires_grad=requires_grad)))
+                  for shape_a, shape_b, shape_c in args_tuple)
+    second_args_tuple = (
```
Comment (imaginary-person) [inline torch/testing/_internal/common_methods_invocations.py:568]: If I combine `second_args_tuple` with `args_tuple` by creating empty dicts, then the `mypy` check fails & asks for a type annotation.

**Comment by imaginary-person** [comment]
Hello @mruberry, please assign yourself as a reviewer for this PR because I'm unable to do so. 

BTW, I added only one sample input for `mm`, as it had only one entry in `method_tests`, but I can add more, if required.

Thanks!

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -3177,16 +3246,21 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
-           dtypesIfCUDA=all_types_and(torch.float16, torch.bfloat16, torch.bool),
            supports_out=False,
            sample_inputs_func=sample_inputs_max_min_reduction_no_dim,),
     OpInfo('mode',
            op=torch.mode,
            dtypes=all_types_and(torch.float16, torch.bfloat16, torch.bool),
-           dtypesIfCPU=all_types_and(torch.float16, torch.bfloat16, torch.bool),
-           dtypesIfCUDA=all_types_and(torch.float16, torch.bfloat16, torch.bool),
            sample_inputs_func=sample_inputs_mode,),
+    OpInfo('mm',
```
Comment (imaginary-person) [inline torch/testing/_internal/common_methods_invocations.py:3255]: TODO: Move this entry above mode's.

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -3224,15 +3289,20 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
            supports_out=False,
            sample_inputs_func=sample_inputs_max_min_reduction_no_dim,),
+    OpInfo('mm',
+           dtypes=floating_and_complex_types_and(torch.half),
+           dtypesIfCPU=all_types_and_complex_and(torch.float16, torch.bfloat16),
+           assert_autodiffed=True,
+           sample_inputs_func=sample_inputs_mm,
+           skips=(
+               # mm does not correctly warn when resizing out= inputs
+               SkipInfo('TestCommon', 'test_out'),
```
Comment (imaginary-person) [inline torch/testing/_internal/common_methods_invocations.py:3301]: This one might have been fixed recently. Will check locally & update.

Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py:3301]: `AssertionError: UserWarning not triggered : Resized a non-empty tensor but did not warn about it.`

**Comment by anjali411** [comment]
Hey @imaginary-person thanks for the PR! sorry this PR went unnoticed. I'll review it by tomorrow!

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -3224,15 +3289,20 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+           )),
     OpInfo('mode',
            op=torch.mode,
            dtypes=all_types_and(torch.float16, torch.bfloat16, torch.bool),
-           dtypesIfCPU=all_types_and(torch.float16, torch.bfloat16, torch.bool),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:3234]: Nice cleanup

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -578,22 +578,85 @@ def sample_inputs_linalg_vector_norm(op_info, device, dtype, requires_grad):
 
     return inputs
 
+def sample_inputs_mm(op_info, device, dtype, requires_grad):
+    args_list = (
+        ((S, M), (M, S)),
+    )
+    inputs = list(SampleInput(make_tensor(first_shape, device, dtype,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:585]: Style: prefer tuple to list (use list only if the iterable has to be mutable)

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -578,22 +578,85 @@ def sample_inputs_linalg_vector_norm(op_info, device, dtype, requires_grad):
...
+                                          requires_grad=requires_grad),
+                              args=(make_tensor(second_shape, device, dtype,
+                                    requires_grad=requires_grad),))
+                  for first_shape, second_shape in args_list)
+    return inputs
+
+def sample_inputs_add(op_info, device, dtype, requires_grad):
+    inputs = []
+    args_list = (
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:594]: cc @kshitij12345 

What's interesting here is that the lhs can require broadcasting to the rhs. How does this pass test_variant_consistency_eager, where the inplace version of the operator is compared to the functional version?

@kshitij12345 is working a PR to support this case now, by the way.

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -578,22 +578,85 @@ def sample_inputs_linalg_vector_norm(op_info, device, dtype, requires_grad):
...
+        inputs.append(SampleInput(make_tensor((S, S, S), device, dtype,
+                                              requires_grad=requires_grad),
+                                  args=(torch.tensor([3.14j], device=device, dtype=dtype,
+                                                     requires_grad=requires_grad),)))
+    return inputs
+
+def sample_inputs_addmm_non_fusible_nodes(op_info, device, dtype, requires_grad):
+    args_tuple = (
+        ((S, M), (S, S), (S, M)),
+        ((1,), (S, S), (S, M)),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:629]: Same question about how this works with the inplace varaint as for mm

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -3224,15 +3289,20 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            op=torch.min,
            variant_test_name='reduction_no_dim',
            dtypes=all_types_and(torch.float16, torch.bfloat16, torch.bool),
-           dtypesIfCPU=all_types_and(torch.float16, torch.bfloat16, torch.bool),
-           dtypesIfCUDA=all_types_and(torch.float16, torch.bfloat16, torch.bool),
            supports_out=False,
            sample_inputs_func=sample_inputs_max_min_reduction_no_dim,),
+    OpInfo('mm',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:3294]: This OpInfo and its sample inputs look good. You may want to separate it into another PR because it could be merged now.

**Comment by mruberry** [comment]
> BTW, I added only one sample input for `mm`, as it had only one entry in `method_tests`, but I can add more, if required.

Nope, this is great for now. Thank you!

**Comment by mruberry** [comment]
Hey @imaginary-person! Thanks for this PR. It's pretty great. I have a question about our testing and how it impacts addmm and add. The mm changes look good, though. If you'd like to separate them we can merge those quickly.

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:594]: lhs broadcasted sample working for inplace variant goes back to our conversation at https://github.com/pytorch/pytorch/pull/54945#discussion_r607039433

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:594]: That's what I thought. We don't have an issue for that, do we? Would you like to submit a PR fixing it now @kshitij12345? If not, I can take a look at it, too. It seems super high priority.

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:629]: @kshitij12345's answer is correct. Let's get that test fixed and then we can sort out these OpInfos.

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:594]: It is taken care of in https://github.com/pytorch/pytorch/pull/53014.

I had drafted a seperate PR for it https://github.com/pytorch/pytorch/pull/55300 (but closed it in favor of #53014)

If it is high-priority, then I can reopen #55300 and get it merged first.
Let me know what you think.

Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py:594]: @kshitij1234, post #53014, would sample inputs that cause an op's inplace variant to resize the base tensor have some kwargs?

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:594]: This PR will update the recommended signature for sample_inputs to `sample_inputs_func(op_info, device, dtype, requires_grad, **kwargs)`. 
With that kwargs can now contain various information from tests if necessary.
For this particular case it will have a flag called `for_inplace_variant` based on which you can update the sample that the test returns.
Eg.
https://github.com/pytorch/pytorch/blob/55076c7a4ac8eb95ed97961b529f20ac3f9f49ef/torch/testing/_internal/common_methods_invocations.py#L1696-L1732

Hope that makes sense. Let me know if you have any questions

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py:629]: @mruberry: Oops, I probably should've noticed something was amiss.

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:629]: @imaginary-person I don't think so. This is a pretty well hidden interaction.

File: torch/testing/_internal/common_methods_invocations.py (hunk 8 above)
Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py]: Opened #55446. Thanks!

**Comment by imaginary-person** [comment]
Hello @heitorschueroff, do you think there's any way to make `mypy` not care about lists or tuples sticking to a particular format? Thanks!

**Comment by kshitij12345** [comment]
@imaginary-person you can either use `# type ignore`
Eg. https://github.com/pytorch/pytorch/blob/10abbb812a47bfabba4905b50dde02a511191522/torch/testing/_internal/common_methods_invocations.py#L712

Or List[SampleInput] and append SampleInput to it. 
Either of them should work. 

With tuples you might have to use `# type ignore` as tuples are supposed to be immutable and hence their type is deduced to be the one at its initialization.
Eg.

https://github.com/pytorch/pytorch/blob/10abbb812a47bfabba4905b50dde02a511191522/torch/testing/_internal/common_methods_invocations.py#L1205-L1213

**Comment by imaginary-person** [comment]
Will fix it locally & then push it, in order to not use CI inordinately.

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -2518,22 +2593,38 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
            skips=(
-               # Skips unsupported bfloat16 check because above support check
-               #   doesn't work on all platforms
-               SkipInfo('TestOpInfo', 'test_unsupported_dtypes', dtypes=(torch.bfloat16,)),
-               # TODO: remove redundant method_tests() entries
-               SkipInfo('TestOpInfo', 'test_duplicate_method_tests')),
+               # Skips unsupported bfloat16 check because cuDNN 8 & below
+               # are buggy. This test works on cuDNN 8.1 & CUDA 11.2
+               SkipInfo('TestOpInfo', 'test_supported_backward', dtypes=[torch.bfloat16], device_type='cuda'),
+           ),
```
Comment (imaginary-person) [inline torch/testing/_internal/common_methods_invocations.py:2612]: Hello @anjali411 & @mruberry, due to the CUDA bug that has been fixed in CUDA 11.2 & cuDNN 8.1, I've enabled `bfloat16` for `addmm` & `mm` on CUDA 11+, but am skipping `test_supported_backward` for it, because CI tests run on CUDA 10.2 & CUDA 11.1 (with older versions of cuDNN). If you'd agree with this approach, maybe it can be done for `addbmm` in #55649 as well. This PR is ready for review. Thank you!

Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py:2612]: Instead of skipping the tests based upon my speculation, I'm running experiments to confirm my hypothesis.

Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py:2612]: `addmm` & `mm`'s `test_supported_backward` are passing on CUDA11 without checking for CUDA SM capability

So, now I know they're not related to `addbmm`, whose corresponding test fails (and will fail) on CUDA 11.1.
I'll unskip these tests for `addmm` & `mm`.

**Review by anjali411 — CHANGES_REQUESTED** [review]
Hey @imaginary-person this PR looks pretty good in terms of both the added opinfos and the cleanup for max, min, mode etc. but as is, it's a little difficult to review this PR. Here's what I suggest:
Split this PR into three PRs: 1. that adds an OpInfo for add 2. that adds another OpInfo for addmm 3. clean up changes

Let me know what you think.

**Comment by imaginary-person** [comment]
Hello @anjali411, I opened #55920 & #55915.
I should probably add the cleanup changes via this PR after all OpInfos would be added, so that removed items won't be re-added. I'll revise this PR to do some more cleanup as well.

File: torch/testing/_internal/common_methods_invocations.py (hunk 9 above)
Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py]: The GPUs used by CircleCI for linux tests are actually SM_52

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py]: Thank you for investigating, @imaginary-person; your attention to detail is impressive.

**Comment by heitorschueroff** [comment]
@imaginary-person What is the current state of this PR? Should it be closed?

**Comment by imaginary-person** [comment]
Hello @heitorschueroff, I apologize for not having updated this PR, but the cleanup seems to have been completed while adding more OpInfos. So, I'll close this PR for now, and would reopen (or open another PR) in the future, if I'd see any room for cleanup. Thank you!
